louislam/uptime-kuma
# respect users specified pgid/puid

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

This PR shows better what I meant in https://github.com/louislam/uptime-kuma/issues/4310. I did not unfortunately have time to properly test all built containers (I see you build for multiple architectures!), but I can confirm that a local container built with `node ./extra/env2arg.js docker buildx build --load -f docker/dockerfile --platform linux/amd64 -t louislam/uptime-kuma:latest --target release .` does work, and I have no reason to suspect it won't work for other platforms. 

## How to test this yourself

1. On the Docker host create a user with a known uid and gid. 
```bash
useradd testuser -u 9999 -g 9999
``` 
3. On the Docker host, create a directory that is owned by the user you just created
```bash 
mkdir testdir && sudo chown testuser:testuser testdir
```
5. Specify the UID and GID in your Docker Compose file like so
```yaml
services:
  uptime-kuma:
    image: louislam/uptime-kuma:latest
    volumes:
      - ./testdir:/app/data
    ports:
      # <Host Port>:<Container Port>
      - 3001:3001
    restart: unless-stopped
    environment:
      - PGID=9999
      - PUID=9999
```
6. `docker compose up -d`
7. `docker logs uptime-kuma-1` should show some log lines showing the node user getting its UID and PID updated.
8. Assuming your container started correctly, connect to its shell and verify that the `node` user now has the correct pid/gid.
```bash
docker exec -it uptime-kuma-1 /bin/bash
# OR you can just run the command directly
docker exec uptime-kuma-1 cat /etc/passwd
```

If you attempt these same steps using the currently published Docker image, you will see that it does not work, as there is nothing handling the PGID/PUID vars. 

## Why like this?
Since we need the UID/GID change to happen when the image is started, we need to execute a shell script every time it starts. Given the desire to not run the main process as root, a shell script that does the change we need, and then immediately drops to the `node` user in order to start the server seemed the easiest way to do this.

Changing the Node uid/gid is no trouble, and is even recommended in [the best practices guide](https://github.com/nodejs/docker-node/blob/main/docs/BestPractices.md#non-root-user) but unfortunately the approach they suggest there won't work for us since we want to allow the user to specify the UID/PID rather than us baking it in. It defaults to 1000/1000 which is what it would have been defaulting to before. A happy accident by the way, since 1000:1000 is the default uid/gid for the first created 'real' user on almost every linux system :P.

## Type of change

Please delete any options that are not relevant.
- Other

## Checklist

- [ ] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [ ] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [ ] My code needed automated testing. I have added them (this is optional task)

Let me know if I can do anything more, I'd love to see this, or a better version of this, merged since that means I can then use it!
